sharktank/sharktank/tools/tuner/candidate_gen.py

```diff
@@ -0,0 +1,1312 @@
+
+    return mlir_module
+
+
+def walk_callback_detect_type(
+    op: ir.Operation, walk_result: OpWalkResult
+) -> ir.WalkResult:
+    if op.name == "linalg.conv_2d_nhwc_hwcf":
+        walk_result.was_interrupted = True
+        walk_result.dispatch_kind = DispatchKind.conv
+        return ir.WalkResult.INTERRUPT
+
+    if op.name == "util.func":
+        func_name = str(op.opview.sym_name)
+        if "batch_matmul_transpose_b" in func_name:
+            walk_result.was_interrupted = True
+            walk_result.dispatch_kind = DispatchKind.batch_mmt
+            return ir.WalkResult.INTERRUPT
+        if "batch_matmul" in func_name:
+            walk_result.was_interrupted = True
+            walk_result.dispatch_kind = DispatchKind.batch_matmul
+            return ir.WalkResult.INTERRUPT
+        if "matmul_transpose_b" in func_name:
+            walk_result.was_interrupted = True
+            walk_result.dispatch_kind = DispatchKind.mmt
+            return ir.WalkResult.INTERRUPT
+        if "matmul_like" in func_name:
+            walk_result.was_interrupted = True
+            walk_result.dispatch_kind = DispatchKind.contraction
+            return ir.WalkResult.INTERRUPT
+    return ir.WalkResult.ADVANCE
+
+
+def walk_mlir_op(mlir_module: ir.Module) -> OpWalkResult:
+    walk_result = OpWalkResult()
+    for op in mlir_module.body.operations:
+        op.walk(
+            lambda op: walk_callback_detect_type(op, walk_result),
+            ir.WalkOrder.POST_ORDER,
+        )
+        if walk_result.was_interrupted:
+            break
+
+    return walk_result
+
+
+def tune(
+    input: str,
+    output: str = "",
+    limit: int = 4096,
+    num_subgroups: int = 4,
+    lhs_dims: str = "mk",
+    rhs_dims: str = "nk",
+    tile_dims: str = "mnk",
+):
+    input_file = str(input)
+
+    if not output:
+        output = get_default_output_dir()
+
+    # Create the directory if it does not exist
+    makedirs(str(output), exist_ok=True)
+
+    tune_logger.debug(f"Output directory {output}")
+    tune_logger.debug(f"Processing {input_file}")
+    mlir_template = read_input_mlir(input_file)
+    mlir_text = "".join(mlir_template)
+
+    mlir_module = parse_mlir(mlir_text)
+    walk_result = walk_mlir_op(mlir_module)
+    assert walk_result.dispatch_kind != None
+
+    # Save the input file as the first candidate.
+    with open(path.join(output, f"0.mlir"), "w") as f:
+        f.write(mlir_text)
+
+    get_shapes_fn: Callable[[list[str]], ProblemSize] | None = None
+    apply_params_fn: (
+        Callable[[ProblemSize, list[str], Configuration], tuple[str, str]] | None
+    ) = None
+    if walk_result.dispatch_kind == DispatchKind.conv:
+        get_shapes_fn = get_shapes_conv
+        apply_params_fn = apply_params_conv
+    elif walk_result.dispatch_kind == DispatchKind.mmt:
+        get_shapes_fn = get_shapes_mmt
+        apply_params_fn = apply_params_mmt
+    elif walk_result.dispatch_kind == DispatchKind.contraction:
```

Comment:
This is a place that could be refactored to make the tuner more modular / easier to extend without needing to keep adding code to this single 1300+ line file.

There are all sorts of ways this could be set up / design patterns this could follow. The first that comes to mind for me right now would be to
1. have a registry of "dispatch kinds"
2. have this code loop over each item in the registry
3. if the item matches the walk result, use the `get_shapes_fn` and `apply_params_fn` from that item

Or... maybe that sort of already exists in the `walk_mlir_op` function - maybe that could set `get_shapes_fn` and `apply_params_fn`?

I'm basically wondering how many places in the script need to be changed to support a new operation. That code would ideally be grouped somehow (e.g. in a function or a subclass of some sort)

Reply:
I added a registry for dispatch kind and did some workaround with op_walk function